### ISSUE TITLE
\/.left / \/.right allocate an extra function

### ISSUE THREAD

**Issue by TomasMikula** [scalaz/scalaz#1611]
`\/.left(a)`/`\/.right(b)` allocate a function only to apply it immediately. Is there a reason why not to change them to methods?

They used to be that way and were changed in https://github.com/scalaz/scalaz/commit/038937f338025f4c749318c2f50aaacad54f842c, for reasons unknown to me.

**Comment by edmundnoble** [comment]
My guess is that @tonymorris wanted to prevent them from taking part in type inference.

**Comment by tonymorris** [comment]
I think they get passed as functions in quite a lot of use-cases.

**Comment by TomasMikula** [comment]
In such use cases, converting the method to a function on demand (`\/.left(_)`) seems to be no worse than `\/.left` returning a function.

**Comment by tonymorris** [comment]
Yeah, no strong opinion from me.

**Comment by tonymorris** [comment]
Maybe have both?

**Comment by edmundnoble** [comment]
I think it's alright to have the function version. The method version should probably be replaced in all cases by `syntax.either`'s postfix methods. The function version gives you the (useful in my code) ability to tell the compiler what type it should expect to be passed as a parameter, which lets you avoid variance issues.

**Comment by TomasMikula** [comment]
> The function version gives you the (useful in my code) ability to tell the compiler what type it should expect to be passed as a parameter, which lets you avoid variance issues.

Can you give an example?

**Comment by edmundnoble** [comment]
Sure.
```scala
sealed trait Thing
case class OneThing(i: Int) extends Thing

def doIt[A](in: Maybe[A]): Maybe[A \/ Thing] = in.map(\/.right[A, Thing] compose OneThing(_))
```
I guess this isn't *that* nice. But I don't see why anyone would use `\/.right` and `\/.left` and not the syntax otherwise.

**Comment by TomasMikula** [comment]
Yeah, this doesn't seem better than `\/.left`/`\/.right` being methods. (You might want to fix your example, though, it doesn't typecheck.)

I guess we could use syntax as smart constructors, though.

**Comment by TomasMikula** [comment]
> Maybe have both?

I guess that would be best.

**Comment by TomasMikula** [comment]
See #1642.

**Comment by TomasMikula** [comment]
Resolved by #1645.
